cmd/prometheus/main.go

```diff
@@ -282,104 +277,193 @@ func main() {
+
+				select {
+				case <-hupReady:
+					break
+				// in case a shutdown is initiated before the hupReady is released
+				case <-cancel:
+					return nil
+				}
 
-	<-dbOpen
+				for {
+					select {
+					case <-hup:
+						if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+							level.Error(logger).Log("msg", "Error reloading config", "err", err)
+						}
+					case rc := <-webHandler.Reload():
+						if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+							level.Error(logger).Log("msg", "Error reloading config", "err", err)
+							rc <- err
+						} else {
+							rc <- nil
+						}
+					case <-cancel:
+						return nil
+					}
+				}
 
-	if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
-		level.Error(logger).Log("msg", "Error loading config", "err", err)
-		os.Exit(1)
+			},
+			func(err error) {
+				close(cancel)
+			},
+		)
 	}
+	{
 
-	defer func() {
-		if err := fanoutStorage.Close(); err != nil {
-			level.Error(logger).Log("msg", "Error stopping storage", "err", err)
-		}
-	}()
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				<-dbOpen
+				if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+					return fmt.Errorf("Error loading config %s", err)
+				}
 
-	// Wait for reload or termination signals.
-	close(hupReady) // Unblock SIGHUP handler.
-
-	// Set web server to ready.
-	webHandler.Ready()
-	level.Info(logger).Log("msg", "Server is ready to receive requests.")
-
-	term := make(chan os.Signal, 1)
-	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
-	select {
-	case <-term:
-		level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
-	case <-webHandler.Quit():
-		level.Warn(logger).Log("msg", "Received termination request via web service, exiting gracefully...")
-	case err := <-errc:
-		level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+				// Wait for reload or termination signals.
+				close(hupReady) // Unblock SIGHUP handler.
+
+				// Set web server to ready.
+				webHandler.Ready()
+				level.Info(logger).Log("msg", "Server is ready to receive requests.")
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				close(cancel)
+			},
+		)
+	}
+	{
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				db, err := tsdb.Open(
+					cfg.localStoragePath,
+					log.With(logger, "component", "tsdb"),
+					prometheus.DefaultRegisterer,
+					&cfg.tsdb,
+				)
+				if err != nil {
+					return fmt.Errorf("Opening storage failed %s", err)
+				}
+				localStorage.Set(db)
+				close(dbOpen)
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				if err := fanoutStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
+				}
+				close(cancel)
+			},
+		)
+	}
+	{
+		// The notifier is a dependency of the rule manager. It has to be
+		// started before and torn down afterwards.
```

Comment:
What is the purpose of this comment? I can't see how this requirement is enforced at the moment.

Reply:
this is a copy and paste form the original implementation and I left it so that I don't forget to ask the same question someone with more context  :) . It is in the TODO list in the PR comment

https://github.com/prometheus/prometheus/blob/dev-2.0/cmd/prometheus/main.go#L334-L340
if both are run in a go routines how is it guaranteed that notifier.Run() will run before ruleManager.Run() ?

comment added with this commit
#7e1b39c

Reply:
IIRC the important part of that commit was not to fix the start order (the order doesn't seem to be deterministic at the moment), but the shutdown order. It's important to first stop the rule manager before the notification handler, as the first will panic if the latter isn't running (anymore). Due to using defer and blocking close methods, that order is ensured.

To not add a regression and bring back the possibility of panics during shutdown, something needs to be done here. Either synchronizing the too shutdown functions by introducing another channel (similar to reloadReady), or by redesigning the implementation of the two actors. I think the first makes more sense in the scope of this PR.

Reply:
somehow I missed this comment and was waiting for some feedback...
anyway , the good news is that  the interrupts are blocking and run in a sequence so I just reordered the blocks and now it is guaranteed that `notifier.Stop()` is called only after `ruleManager.Stop()` exited.

here is the Actors interrupt implementation
https://github.com/oklog/oklog/blob/v0.2.2/pkg/group/group.go#L46-L48

Reply:
feel free to change any comments format/text if something doesn't sound right.